File: src/download.ts
import type { Bot } from "grammy"

// ── File download helper ────────────────────────────────────────────────────

export type AttachedFile = { mime: string; filename: string; dataUrl: string }

export async function downloadTelegramFile(
  bot: Bot,
  fileId: string,
  mime: string,
  filename: string,
): Promise<AttachedFile | undefined> {
  try {
    const file = await bot.api.getFile(fileId)
    if (!file.file_path) return undefined
    // Use grammy's built-in file URL construction to avoid leaking token in error messages
    const url = `https://api.telegram.org/file/bot${bot.token}/${file.file_path}`
    const res = await fetch(url)
    if (!res.ok) return undefined
    const buf = Buffer.from(await res.arrayBuffer())
    const dataUrl = `data:${mime};base64,${buf.toString("base64")}`
    return { mime, filename, dataUrl }
  } catch (e) {
    console.error("File download error:", (e as Error).message)
    return undefined
  }
}

File: src/download.rs
use crate::telegram::TelegramClient;
use base64::Engine;

pub struct AttachedFile {
    pub mime: String,
    pub filename: String,
    pub data_url: String,
}

pub async fn download_telegram_file(
    client: &TelegramClient,
    file_id: &str,
    mime: &str,
    filename: &str,
) -> Option<AttachedFile> {
    let file = client.get_file(file_id).await.ok()?;
    let file_path = file.file_path?;
    let bytes = client.download_file_bytes(&file_path).await.ok()?;
    let b64 = base64::engine::general_purpose::STANDARD.encode(&bytes);
    let data_url = format!("data:{};base64,{}", mime, b64);
    Some(AttachedFile {
        mime: mime.to_string(),
        filename: filename.to_string(),
        data_url,
    })
}
